File: jan2015/p1/CowRouting.cpp
//////////
// USACO UNOFFICIAL SOLUTION CONTEST 2 PROBLEM 1
// BRANDON LIN
// C++
/////////
#include<iostream>
#include<vector>
#include<cstdio>
#include<algorithm>

using namespace std;
int main(){
  int city1, city2, length;
  freopen("cowroute.in", "r", stdin);
  freopen("cowroute.out", "w", stdout);
  cin >> city1 >> city2 >> length;
  int cost, numCities, min = 1001;
  vector<int> cities(500,1);
  bool value = false;
  for(int i = 0; i < length; i++){
    cin >> cost >> numCities;
    for(int j = 0; j < numCities; j++){
      cin >> cities[j];
    }
    for(int k = 0; k < numCities; k++){
      if(cities[k] == city1){
	value = true;
      }else if(cities[k] == city2 && !value){
	break;
      }else if(cities[k] == city2 && value && cost < min){
	min = cost;
	break;
      }else if(cities[k] == city2 && value && cost >= min){
	break;
      }
    }
    value = false;
  }
  if( min == 1001){
    min = -1;
  }
  cout << min << "\n";
  return 0;
}

File: 2014-2015/2-jan2015/p1/CowRouting.py
#########
# USACO CONTEST 2 PROBLEM 1
# SOLUTION BY BRANDON LIN
# PYTHON 2.7.6
#########
import sys

def readin(): #Read in the file
    f = open("cowroute.in",'r')
    s = f.read().split("\n")
    f.close()
    return s

def find(list,el): # A modified linear search for lists
    if el in list:
        return list.index(el)
    else:
        return -1
l = readin()
min = sys.maxint
firstline = l[0].split()
count = 2
while count < len(l):
    if int(l[count - 1].split()[0]) < min and find(l[count].split(),firstline[0]) != -1 and find(l[count].split(),firstline[1]) != -1 and find(l[count].split(),firstline[0]) < find(l[count].split(),firstline[1]): # Check to see if the two cities occur in the same order
        min = int(l[count - 1].split()[0]) 
    count += 2

if min == sys.maxint: # If we haven't touched the minimum, make it -1
    min = -1

print str(min)
    
g = open("cowroute.out",'w')
g.write(str(min) + "\n")
g.close()
